pkg/ocibundle/tools/overlay_linux.go

```diff
@@ -30,22 +35,47 @@ func CreateOverlay(bundlePath string) error {
 		}
 	}()
 
-	err = syscall.Mount(overlayDir, overlayDir, "", syscall.MS_BIND, "")
-	if err != nil {
-		return fmt.Errorf("failed to bind %s: %s", overlayDir, err)
+	return MountOverlay(bundlePath, OverlaySet{WritableLoc: overlayDir})
+}
+
+func MountOverlay(bundlePath string, ovs OverlaySet) error {
+	var err error
+
+	locsToBind := ovs.ReadonlyLocs
+	if len(ovs.WritableLoc) > 0 {
+		// Check if writable overlay dir already exists; if it doesn't, try to create it.
+		if err = ensureOverlayDir(ovs.WritableLoc, true, 0o755); err != nil {
+			return err
+		}
+
+		locsToBind = append(locsToBind, ovs.WritableLoc)
 	}
-	// best effort to cleanup mount
-	defer func() {
-		if err != nil {
-			syscall.Unmount(overlayDir, syscall.MNT_DETACH)
+
+	// Try to do initial bind-mounts
+	for _, d := range locsToBind {
+		if err = ensureOverlayDir(d, false, 0); err != nil {
+			return fmt.Errorf("error accessing directory %s: %s", d, err)
 		}
-	}()
 
-	if err = syscall.Mount("", overlayDir, "", syscall.MS_REMOUNT|syscall.MS_BIND, ""); err != nil {
-		return fmt.Errorf("failed to remount %s: %s", overlayDir, err)
+		if err = syscall.Mount(d, d, "", syscall.MS_BIND, ""); err != nil {
+			return fmt.Errorf("failed to bind %s: %s", d, err)
+		}
+
+		// best effort to cleanup mount
+		defer func() {
+			if err != nil {
+				syscall.Unmount(d, syscall.MNT_DETACH)
+			}
+		}()
+
+		// Try to perform remount
+		if err = syscall.Mount("", d, "", syscall.MS_REMOUNT|syscall.MS_BIND, ""); err != nil {
+			return fmt.Errorf("failed to remount %s: %s", d, err)
+		}
 	}
 
-	err = prepareOverlay(bundlePath, overlayDir)
+	// Prepare internal structure of overlay dir
```

Comment:
This comment is a little misleading, as `prepareOverlay` also performs the mount.